### ISSUE TITLE
Why `preserve_rng_state=False` in activation checkpointing

### ISSUE THREAD

**Issue by awgu** [pytorch/torchtitan#1323]
Why does torchtitan set `preserve_rng_state=False` for activation checkpointing? E.g.:
https://github.com/pytorch/torchtitan/blob/f4048f8e1b36827156c4dc861c9680333a8542f9/torchtitan/models/llama3/infra/parallelize.py#L238

**Comment by tianyu-l** [comment]
@soulitzer for any context / suggestions

**Comment by soulitzer** [comment]
It is surprising to see `preserve_rng_state=False`. Although you remove the overhead of stashing/restoring RNG state, if any operations use randomness, then you can end up with incorrect gradients.

**Comment by tianyu-l** [comment]
@soulitzer I see.

I believe Llama is not using randomness -- for any model whose model code doesn't use randomness, it should be fine either way? If a model is using randomness then it's recommended to use `preserve_rng_state=True`. Is this the right conclusion?

Then I wonder why we have this option at all -- why don't we only support the `True` case for correctness. Is it because some model doesn't care about incorrect gradients from recomputation, so we can save the overhead?

**Comment by soulitzer** [comment]
> If a model is using randomness then it's recommended to use preserve_rng_state=True. Is this the right conclusion?

Yeah

> I believe Llama is not using randomness

Are we using dropout in sdpa btw? (This reminds me that for SAC I might be able to stop treating it as a random op if dropout_p is set to 0)

> so we can save the overhead?

We might've checked this at some point to see if the overhead was minor enough to just not care about, I don't think that was the case.

**Comment by awgu** [comment]
The extra overhead is mainly some extra CPU overhead?

**Comment by tianyu-l** [comment]
@soulitzer 
Llama is not using dropout in sdpa afaik.

> We might've checked this at some point to see if the overhead was minor enough to just not care about, I don't think that was the case.

Then why don't we always set it to True?

**Comment by soulitzer** [comment]
> Then why don't we always set it to True?

@tianyu-l Sorry I worded that kind of badly lol, what I mean to say is the overhead didn't seem minor enough to ignore :P

> The extra overhead is mainly some extra CPU overhead?

@awgu Not sure, we just looked at the runtime and didn't try to stare too hard at it. Maybe worth checking again 🤔

**Comment by sippycoder** [comment]
I just experimented with MoE layers in my experiments. Activation checkpointing does not work without `preserve_rng_state=True`. The router is pretty sensitive to small deviations. There are a number of reasons that can recompute the router differently in backward pass, 

- bfloat16/float16 casts
- softmax is extremely sensitive
- CUDA math kernels that are not bit-exact: `flash_attn_varlen_func`, grouped-GEMM, silu, etc. use CUDA atomics/reductions whose result depends on thread scheduling. They return numerically identical values only up to the last 1-2 umps; two successive launches of the same input generally differ by up to 1 ump or so.

At least for MoE model, `preserve_rng_state` needs to be `True`.

`torchtitan` cleverly skips checkpointing `moe.router.gate` altogether. So does `Megatron`. I wouldn't be surprised if `preserve_rng_state=True` is not enough for checkpointing the router correctly. I will run an experiment and get back.

**Comment by tianyu-l** [comment]
@sippycoder 
Thanks for the note!

> torchtitan cleverly skips checkpointing moe.router.gate altogether.

By this you mean `torchtitan` always recompute `moe.router.gate`?
For SAC I think it's because of https://github.com/pytorch/torchtitan/pull/1380

> bfloat16/float16 casts
softmax is extremely sensitive

For this do you think we should consider change `moe.router.gate` to compute in fp32? Currently I believe it's in bf16, governed by FSDP2 mixed precision behavior.

**Comment by sippycoder** [comment]
> By this you mean torchtitan always recompute moe.router.gate?

Nope. I mean, the `router` computations are *skipped*. They try to avoid dealing with the randomness in the re-computations by avoiding the recompute for  `moe.router.gate` and all the fqn's that match same shape as `moe.router.gate` (a side effect).

> For this do you think we should consider change moe.router.gate to compute in fp32? Currently I believe it's in bf16, governed by FSDP2 mixed precision behavior.

It's already done in `torch.float32` I believe. Even with that, we can still to avoid the re-computation of routing because it's highly sensitive.

**Comment by tianyu-l** [comment]
@sippycoder 
> They try to avoid dealing with the randomness in the re-computations by **avoiding the recompute for moe.router.gate** and all the fqn's that match same shape as moe.router.gate (a side effect).

What makes you believe that? The code says prefere / force RECOMPUTE moe.router.gate.
https://github.com/pytorch/torchtitan/pull/1380/files#diff-5a556cc02e3e89c77936577320ae0d6c7382ad291ee65d32eea08291dc3618e1R299

> It's already done in torch.float32 I believe.

What makes you believe that?

**Comment by sippycoder** [comment]
I went back and rechecked the code.

> What makes you believe that? The code says prefere / force RECOMPUTE moe.router.gate.

Looks like they ARE forcing the recomputing of router gate. I'm surprised, given the sensitivity of it. 

`moe.router.gate` is also not in float32, only the scores coming out of the gate are cast to float32. This is fine as long as recomputations are preserving rng state imo.

**Comment by tianyu-l** [comment]
@soulitzer 

Let's force save of router, instead of force recompute, due to stability concerns.

How do we do this with full AC?

**Comment by tianyu-l** [comment]
@sippycoder 
How is router save / recomputation related to `preserve_rng_state=True`?
IIUC `router.gate` is a `nn.Linear` which doesn't have randomness.

**Comment by sippycoder** [comment]
For `router.gate`, `nn.Linear` should be fine. It's the `F.softmax` that is sensitive. Even 1-ULP rounding difference can give us a difference topK and that's a huge issue for backward pass. We should either make them deterministic or avoid checkpointing.

**Comment by tianyu-l** [comment]
@sippycoder 
Thanks! So would your suggestions be
- `router.gate` can be done in bf16. Recomputation on it is fine even with `preserve_rng_state=False`.
- `softmax` has to be checkpointed, or have deterministic results after recomputation.

My questions would be
- Do you know that `preserve_rng_state=True` will definitely give deterministic results for `softmax`?
- How about [`sigmoid`](https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/llama4/model/moe.py#L165)? Is the results also very sensitive?

**Comment by soulitzer** [comment]
@sippycoder 
Maybe I'm missing something, but I guess don't understand yet why preserve_rng_state is related to the small observed difference. If the rng state isn't preserved and there is randomness involved, I would expect there to be large differences rather than small ones? Separately there's also https://docs.pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html

@tianyu-l 

> Let's force save of router, instead of force recompute, due to stability concerns.

How concerned are we about backward compatibility wrt TorchTitan configs, someone using old config with new TorchTitan version? If not, maybe one thing we could do is just make the config more general to allow specifying either save/recompute "router.gate": "save".

> How do we do this with full AC?

This isn't possible to do in full AC! We need SAC to enable AC to save intermediates.

**Comment by tianyu-l** [comment]
@soulitzer 
I think it's fine to break BC in this case.

Since moe router is the only thing of interest right now, I wonder if the change could be just change `per_op_sac_force_recompute_mm_shapes_by_fqns` to `per_op_sac_force_save_mm_shapes_by_fqns` and `PREFER_RECOMPUTE` to `MUST_SAVE`

**Comment by soulitzer** [comment]
I see, that sounds okay to me. Athough I'm still not sure exactly what is going on with preserve_rng_state and why it would cause small differences.

**Comment by tianyu-l** [comment]
I think these may be separate issues. Also i realized my tentative conclusion was
> Thanks! So would your suggestions be
> - `router.gate` can be done in bf16. Recomputation on it is fine even with `preserve_rng_state=False`.
> - `softmax` has to be checkpointed, or have deterministic results after recomputation.

If that's the case we should save softmax, instead of router.gate. @sippycoder could you confirm the details?

**Comment by soulitzer** [comment]
Oh actually I think what @sippycoder  is saying is that it is not related to `preserve_rng_state`, but you have certain math kernels that can differ:

> CUDA math kernels that are not bit-exact: flash_attn_varlen_func, grouped-GEMM, silu, etc. use CUDA atomics/reductions whose result depends on thread scheduling. They return numerically identical values only up to the last 1-2 umps; two successive launches of the same input generally differ by up to 1 ump or so.

Although the main (only?) place where this matters is computing the top-k (so even if the the non-determinism originated from the cuda math kernels, saving the bf16 router.gate will be mostly okay too), if it is obvious which kernels fall into the "CUDA math kernels" case, perhaps its more accurate to save just their outputs instead?

**Comment by wwwjn** [comment]
@soulitzer  is going to benchmark the overhead of setting default to True

**Comment by soulitzer** [comment]
I measured the overhead of setting default to True to be a constant 30-50us on 8gpu node per AC region. For comparison, AC already has a 350us constant overhead. Overhead scales linearly when inputs are on more distinct devices, but that is rare. This should matter less for the Titan use case anyway because AC regions are fairly large (transformer blocks). It is possible the constant overhead will matter more when only several ops are being AC'd. Created PR here to flip the switch https://github.com/pytorch/torchtitan/pull/2380.
